server/matchmaker/algorithm.py

```diff
@@ -87,42 +93,112 @@ def _propose(self, search: Search, preferred: Search):
+                if search != newbie
+                and not search in self.matches
+                and search.is_single_party()
+                and search.has_no_top_player()
+                ),
+                default_if_no_available_opponent
+            )
+            if opponent is not default_if_no_available_opponent:
+                self._match(newbie, opponent)
+
+        return self.matches
+
+
+class Matchmaker(object):
+    def __init__(self, searches: List[Search]):
+        self.searches = searches
+        self.matches: Dict[Search, Search] = {}
+
+    def find(self) -> List[Match]:
+        self.matches.update(StableMarriage(self.searches).find())
 
-def _rank_all(searches: List[Search]) -> Dict[Search, List[Search]]:
-    """ Returns searches with best quality for each search.
+        remaining_searches = [
+                search for search in self.searches 
+                if search not in self.matches
+        ]
+        self.matches.update(RandomlyMatchNewbies(remaining_searches).find())
 
-    Note that the highest quality searches come at the end of the list so that
-    it can be used as a stack with .pop().
-    """
-    return {
-        search: sorted(
-            _rank_partners(
-                search, filter(lambda s: s is not search, searches)
-            ),
+        return self._remove_duplicates()
+
+    def _remove_duplicates(self) -> List[Match]:
+        matches_set: Set[Match] = set()
+        for s1, s2 in self.matches.items():
+            if (s1, s2) in matches_set or (s2, s1) in matches_set:
+                continue
+            matches_set.add((s1, s2))
+        return list(matches_set)
+
+
+@with_logger
+class _MatchingGraph:
+    @staticmethod
+    def build_sparse(searches: List[Search]) -> Dict[Search, List[Search]]:
+        """ A graph in adjacency list representation, whose nodes are the searches
+        and whose edges are the top few possible matchings for each node.
+
+        Note that the highest quality searches come at the end of the list so that
+        it can be used as a stack with .pop().
+        """
+        return {
+            search: sorted(
+                _MatchingGraph._get_top_matches(
+                    search, filter(lambda s: s is not search, searches)
+                ),
+                key=lambda other: search.quality_with(other)
+            )
+            for search in searches
+        }
+
+
+    @staticmethod
+    def _get_top_matches(search: Search, others: Iterable[Search]) -> List[Search]:
+        def is_possible_match(other: Search) -> bool:
+            quality_log_string = (
+                f"Quality between {search} and {other}: {search.quality_with(other)}"
+                + f" thresholds: [{search.match_threshold}, {other.match_threshold}]."
+            )
+
+            if search.matches_with(other):
+                _MatchingGraph._logger.debug(
+                    quality_log_string + f" Will be considered during matchmaking."
```

Comment:
In general string formatting is better than concatenation because it will require less copying. If you want to go down the concatenation route, then the string doesn't need to be an f-string.
```suggestion
                    f"{quality_log_string} Will be considered during matchmaking."
```

Reply:
Oh, that's nice, thank you!